File: Scala/src/main/scala/TextbookTrack/Chapter08/BA8E.scala
package TextbookTrack.Chapter08

object BA8E {
  import scala.annotation.tailrec

  final case class Cluster(id: Int, content: Vector[Int]) {
    val size: Int = content.length
  }

  final case class IndexPair(i: Int, j: Int) {
    require(i <= j, s"Index $i should be less than index $j.")

    def contains(k: Int): Boolean = (k == i) || (k == j)
  }

  class DistanceMatrix(distances: Map[IndexPair, Double]) {
    def apply(key: IndexPair): Double = distances(key)

    private def clusterIds: Set[Int] = distances.keysIterator.flatMap{ case IndexPair(i, j) => Set(i, j) }.toSet

    def closestClusters(): IndexPair = distances.minBy{ case (_, distance) => distance }._1

    def update(cluster1: Cluster, cluster2: Cluster, nextId: Int): DistanceMatrix = {
      val updatedDistances: Map[IndexPair, Double] =
        clusterIds.map{
          id =>
            val pair1 = IndexPair(math.min(id, cluster1.id), math.max(id, cluster1.id))
            val pair2 = IndexPair(math.min(id, cluster2.id), math.max(id, cluster2.id))
            val num: Double =
              distances.getOrElse(pair1, 0.0) * cluster1.size + distances.getOrElse(pair2, 0.0) * cluster2.size
            IndexPair(id, nextId) -> num / (cluster1.size + cluster2.size)
        }.toMap ++ distances

      val reducedDistances: Map[IndexPair, Double] =
        updatedDistances.filter{ case (key, _) => !key.contains(cluster1.id) && !key.contains(cluster2.id)}
      new DistanceMatrix(reducedDistances)
    }
  }

  private def convertToDoubleList(line: String): List[Double] = line.split(" ").map(_.toDouble).toList

  def readDistanceMatrix(reader: Iterator[String], n: Int): DistanceMatrix = {
    val adjacencyList: Map[IndexPair, Double] =
      reader
        .map(convertToDoubleList)
        .take(n)
        .zipWithIndex
        .flatMap{ case (row, ix) => row.zipWithIndex.drop(ix + 1).map{ case (dist, jy) => (IndexPair(ix, jy), dist) } }
        .toMap
    new DistanceMatrix(adjacencyList)
  }

  def runHierarchicalClustering(distances: DistanceMatrix, n: Int): List[Cluster] = {
    @tailrec
    def loop(clusters: Map[Int, Cluster],
             distances: DistanceMatrix,
             newClusters: List[Cluster],
             nextId: Int): List[Cluster] = {
      if (clusters.size <= 1) newClusters.reverse
      else {
        val IndexPair(id1, id2): IndexPair = distances.closestClusters()
        val cluster1: Cluster = clusters(id1)
        val cluster2: Cluster = clusters(id2)
        val nextCluster = Cluster(id = nextId, content = cluster1.content ++ cluster2.content)
        val updatedClusters: Map[Int, Cluster] = clusters -- Set(id1, id2) + (nextId -> nextCluster)
        loop(updatedClusters, distances.update(cluster1, cluster2, nextId), nextCluster :: newClusters, nextId + 1)
      }
    }

    val initialClusters: Map[Int, Cluster] = (0 until n).map(k => k -> Cluster(k, Vector(k))).toMap
    loop(initialClusters, distances, Nil, n)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val distances: DistanceMatrix = readDistanceMatrix(reader, n)
    val result: List[Cluster] = runHierarchicalClustering(distances, n)
    result.foreach{ case Cluster(_, content) => println(content.map(_ + 1).mkString(" ")) }
  }
}

File: Python/textbook_track/chapter08/ba8e.py
import sys

from collections import namedtuple
from itertools import count

Cluster = namedtuple('Cluster', ['id', 'content'])


class DistanceMatrix:
    def __init__(self, distances):
        self._distances = distances

    def __getitem__(self, key):
        i, j = key
        if i < j:
            return self._distances[(i, j)]
        if i == j:
            return 0
        return self._distances[(j, i)]

    def __setitem__(self, key, value):
        i, j = key
        if i < j:
            self._distances[(i, j)] = value
        elif i > j:
            self._distances[(j, i)] = value

    def _current_cluster_ids(self):
        ids = set()
        for key in self._distances.keys():
            ids.update(key)
        return frozenset(ids)

    def closest_clusters(self):
        return min(self._distances, key=self._distances.get)

    def update(self, cluster_1, cluster_2, next_id):
        cluster_ids = self._current_cluster_ids()
        for cluster_id in cluster_ids:
            size_1 = len(cluster_1.content)
            size_2 = len(cluster_2.content)
            num = self[(cluster_id, cluster_1.id)] * size_1 + self[(cluster_id, cluster_2.id)] * size_2
            self[(cluster_id, next_id)] = num / (size_1 + size_2)

        self._distances = {key: dist for key, dist in self._distances.items()
                           if cluster_1.id not in key and cluster_2.id not in key}


def convert_to_floatlist(line):
    return [float(elem) for elem in line.split()]


def read_distance_matrix(reader, n):
    distances = {}
    for ix in range(n):
        row = convert_to_floatlist(next(reader))
        for jy in range(ix+1, n):
            distances[(ix, jy)] = row[jy]
    return DistanceMatrix(distances)


def run_hierarchical_clustering(distances, n):
    id_generator = count(n)
    clusters = {k: Cluster(id=k, content=(k,)) for k in range(n)}
    new_clusters = []
    while len(clusters) > 1:
        id_1, id_2 = distances.closest_clusters()
        cluster_1 = clusters.pop(id_1)
        cluster_2 = clusters.pop(id_2)
        next_id = next(id_generator)
        next_cluster = Cluster(id=next_id, content=cluster_1.content+cluster_2.content)
        clusters[next_id] = next_cluster
        new_clusters.append(next_cluster)
        distances.update(cluster_1, cluster_2, next_id)
    return new_clusters


def main():
    reader = sys.stdin
    n = int(next(reader))
    distances = read_distance_matrix(reader, n)
    result = run_hierarchical_clustering(distances, n)
    for cluster in result:
        print(' '.join(map(lambda x: str(x + 1), cluster.content)))


if __name__ == '__main__':
    main()
